server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java

```diff
@@ -0,0 +1,194 @@
+         // and if we get values complete the listener future. In case of exception
+         // completeExceptionally
+         return CompletionStages.handleAndCompose(pollAllKeys(listMultimap, filterKeys), (v, t2) -> {
+            // If second poll fails, remove listener and complete exceptionally
+            if (t2 != null) {
+               cache.removeListenerAsync(pubSubListener);
+               pubSubFuture.completeExceptionally(t);
+               return pubSubFuture;
+            }
+            if (v != null) {
+               // poll returns values, complete the listener future ...
+               pubSubFuture.complete(v);
+            } else {
+               // ... else wait for values with timeout if required
+               pubSubListener.startTimer(timeout);
+            }
+            pubSubFuture.thenRun(() -> {
+               pubSubListener.deleteTimer();
+               cache.removeListenerAsync(pubSubListener);
+            });
+            return pubSubFuture;
+         });
+      });
+   }
+
+   private CompletionStage<Collection<byte[]>> pollAllKeys(EmbeddedMultimapListCache<byte[], byte[]> listMultimap,
+         List<byte[]> filterKeys) {
+      var pollStage = pollKeyValue(listMultimap, filterKeys.get(0));
+      for (int i = 1; i < filterKeys.size(); ++i) {
+         var keyChannel = filterKeys.get(i);
+         pollStage = pollStage.thenCompose(
+               v -> (v == null || v.isEmpty())
+                     ? pollKeyValue(listMultimap, keyChannel)
+                     : CompletableFuture.completedFuture(v));
+      }
+      return pollStage;
+   }
+
+   @Listener(clustered = true)
+   public static class PubSubListener {
+      private final Channel channel;
+      ScheduledFuture<?> scheduledTimer;
+      private CompletableFuture<Collection<byte[]>> future = new CompletableFuture<>();
+
+      public CompletableFuture<Collection<byte[]>> getFuture() {
+         return future;
+      }
+
+      public PubSubListener(Channel channel) {
+         this.channel = channel;
+      }
+
+      public void startTimer(long timeout) {
+         deleteTimer();
+         scheduledTimer = (timeout > 0) ? channel.eventLoop().schedule(() -> {
+            future.complete(null);
+         }, timeout, TimeUnit.MILLISECONDS) : null;
+      }
+
+      public void deleteTimer() {
+         if (scheduledTimer != null)
+            scheduledTimer.cancel(true);
+         scheduledTimer = null;
+      }
+
+      @CacheEntryCreated
+      public CompletionStage<Void> onEvent(CacheEntryEvent<Object, Object> entryEvent) {
+         try {
+            if (entryEvent.getValue() instanceof ListBucket) {
+               byte[] key = unwrapKey(entryEvent.getKey());
+               var bucket = (ListBucket<byte[]>) entryEvent.getValue();
+               byte[] value = bucket == null ? null : bucket.toDeque().peek();
+               future.complete(Arrays.asList(key, value));
```

Comment:
Doesn't this need to pop the value it saw? It seems to only just return it and doesn't pop afai can see.